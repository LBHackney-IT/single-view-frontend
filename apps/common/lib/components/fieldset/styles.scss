@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-fieldset/fieldset";

.govuk-fieldset > .govuk-visually-hidden + .mtfh-fieldset__content {
  &,
  & > *:first-child {
    margin-top: 0;
  }
}

.mtfh-fieldset {
  &--indent > .mtfh-fieldset__content {
    padding-left: govuk-spacing(3);
    border-left: $govuk-border-width-form-group-error solid lbh-colour("lbh-grey-2");
  }

  &--error > .mtfh-fieldset__content {
    padding-left: govuk-spacing(3);
    border-left: $govuk-border-width-form-group-error solid lbh-colour("lbh-error");
  }

  &__content {
    &:empty {
      margin-top: 0;
    }
  }
}
