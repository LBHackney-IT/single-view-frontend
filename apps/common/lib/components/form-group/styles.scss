@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
