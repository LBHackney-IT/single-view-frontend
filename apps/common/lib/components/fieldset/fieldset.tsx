import React, {
  ComponentPropsWithoutRef,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
} from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";
import "./styles.scss";

export interface FieldsetProps extends ComponentPropsWithoutRef<"fieldset"> {
  heading: string | ReactElement<ComponentPropsWithoutRef<"h1">>;
  variant?: "base" | "small" | "medium" | "large" | "xlarge" | "hidden";
  indent?: boolean;
  error?: boolean | string;
  override?: number;
}

export const Fieldset = forwardRef<HTMLFieldSetElement, FieldsetProps>(function Fieldset(
  {
    variant = "base",
    indent = false,
    error,
    heading,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  return (
    <fieldset
      ref={ref}
      className={cn(
        "govuk-fieldset",
        "lbh-fieldset",
        {
          "mtfh-fieldset--indent": indent,
          "mtfh-fieldset--error": !!error,
        },
        widthOverrides(override),
        className,
      )}
      {...props}
    >
      <legend
        className={cn("govuk-fieldset__legend", {
          "govuk-fieldset__legend--s": variant === "small",
          "govuk-fieldset__legend--m": variant === "medium",
          "govuk-fieldset__legend--l": variant === "large",
          "govuk-fieldset__legend--xl": variant === "xlarge",
          "govuk-visually-hidden": variant === "hidden",
        })}
      >
        {typeof heading !== "string"
          ? isValidElement(heading) &&
            cloneElement(heading, {
              className: cn("govuk-fieldset__heading", heading.props.className),
            })
          : heading}
        {error && <div className="govuk-visually-hidden"> {error}</div>}
      </legend>
      <div className={cn("mtfh-fieldset__content")}>
        {error && (
          <span className="govuk-error-message lbh-error-message" aria-hidden="true">
            {error}
          </span>
        )}
        {children}
      </div>
    </fieldset>
  );
});
