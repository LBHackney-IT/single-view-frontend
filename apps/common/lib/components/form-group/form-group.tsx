import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
