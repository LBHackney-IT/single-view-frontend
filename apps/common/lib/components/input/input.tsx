import React, { ComponentPropsWithoutRef, forwardRef } from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import "./styles.scss";

export interface InputProps extends ComponentPropsWithoutRef<"input"> {
  error?: boolean;
  override?: number;
}

export const Input = forwardRef<HTMLInputElement, InputProps>(function Input(
  { error, className, override, ...props },
  ref,
) {
  const inputClasses = classNames(
    "govuk-input",
    "lbh-input",
    {
      "govuk-input--error": error,
    },
    widthOverrides(override),
    className,
  );

  return <input ref={ref} className={inputClasses} {...props} />;
});
