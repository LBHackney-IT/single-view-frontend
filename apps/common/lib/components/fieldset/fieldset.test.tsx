import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { FormGroup } from "../form-group";
import { Input } from "../input";
import { Fieldset } from "./fieldset";

test("it renders correctly", async () => {
  const { container } = render(
    <Fieldset heading="Group name">
      <FormGroup id="test" label="Test" name="test">
        <Input />
      </FormGroup>
    </Fieldset>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly with an h tag", async () => {
  const { container } = render(
    <Fieldset heading={<h1>Heading</h1>}>
      <FormGroup id="test" label="Test" name="test">
        <Input />
      </FormGroup>
    </Fieldset>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it applies the correct variant styling", () => {
  const { container } = render(
    <Fieldset variant="large" heading={<h1>Heading</h1>}>
      <FormGroup id="test" label="Test" name="test">
        <Input />
      </FormGroup>
    </Fieldset>,
  );
  expect(container).toMatchSnapshot();
});

test("it applies an error", async () => {
  const { container } = render(
    <Fieldset variant="large" heading="heading" error="This is an error">
      <FormGroup id="test" label="Test" name="test">
        <Input />
      </FormGroup>
    </Fieldset>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});
