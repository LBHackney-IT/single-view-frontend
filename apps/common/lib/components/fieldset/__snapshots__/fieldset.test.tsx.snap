// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it applies an error 1`] = `
<div>
  <fieldset
    class="govuk-fieldset lbh-fieldset mtfh-fieldset--error"
  >
    <legend
      class="govuk-fieldset__legend govuk-fieldset__legend--l"
    >
      heading
      <div
        class="govuk-visually-hidden"
      >
         
        This is an error
      </div>
    </legend>
    <div
      class="mtfh-fieldset__content"
    >
      <span
        aria-hidden="true"
        class="govuk-error-message lbh-error-message"
      >
        This is an error
      </span>
      <div
        class="govuk-form-group lbh-form-group"
        id="test"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field"
        >
          Test
          
        </label>
        <input
          class="govuk-input lbh-input"
          id="test-field"
          name="test"
        />
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`it applies the correct variant styling 1`] = `
<div>
  <fieldset
    class="govuk-fieldset lbh-fieldset"
  >
    <legend
      class="govuk-fieldset__legend govuk-fieldset__legend--l"
    >
      <h1
        class="govuk-fieldset__heading"
      >
        Heading
      </h1>
    </legend>
    <div
      class="mtfh-fieldset__content"
    >
      <div
        class="govuk-form-group lbh-form-group"
        id="test"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field"
        >
          Test
          
        </label>
        <input
          class="govuk-input lbh-input"
          id="test-field"
          name="test"
        />
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <fieldset
    class="govuk-fieldset lbh-fieldset"
  >
    <legend
      class="govuk-fieldset__legend"
    >
      Group name
    </legend>
    <div
      class="mtfh-fieldset__content"
    >
      <div
        class="govuk-form-group lbh-form-group"
        id="test"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field"
        >
          Test
          
        </label>
        <input
          class="govuk-input lbh-input"
          id="test-field"
          name="test"
        />
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`it renders correctly with an h tag 1`] = `
<div>
  <fieldset
    class="govuk-fieldset lbh-fieldset"
  >
    <legend
      class="govuk-fieldset__legend"
    >
      <h1
        class="govuk-fieldset__heading"
      >
        Heading
      </h1>
    </legend>
    <div
      class="mtfh-fieldset__content"
    >
      <div
        class="govuk-form-group lbh-form-group"
        id="test"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field"
        >
          Test
          
        </label>
        <input
          class="govuk-input lbh-input"
          id="test-field"
          name="test"
        />
      </div>
    </div>
  </fieldset>
</div>
`;
